### ISSUE TITLE
Implementation choice for generating code comments

### ISSUE THREAD

**Issue by palmerst** [JacquesCarette/Drasil#444]
We want a choice for generating code comments (and sub-choices for specifics on what should be commented).  The commenting features of GOOL will need an overhaul to do this right.  Currently comments in GOOL are implemented as part of the Statement syntax, so they can only be put with/between statements.  There is no way to make comment blocks for modules/classes/functions.

**Comment by bmaclach** [comment]
On the GOOL side of things, there are currently two ways to do comments. `comment` for single-line comments, represented as Statements, and `addComments` for adding a single-line comment above and below a Body of code, as a sort of delimiter with a short description. We can easily add a `blockComment` function. I am thinking this function should reside in a new typeclass (thinking from the perspective of finally tagless GOOL). Then we can add to the `MethodSym`, `ClassSym`, and `ModuleSym` typeclasses functions like `methodWithComments`, `classWithComments`, `moduleWithComments` which are parameterized by a `BlockComment` in addition to all of their regular parameters. Our renderers would then display this comment directly above the corresponding method/class/module. The reason for not adding `blockComment` to the `BlockSym` typeclass is that we would then need to parameterize functions for creating classes and modules with a Block, which could easily be abused (by putting actual code in the Block instead of a block comment).

On the Drasil side of things, our `Choices` data structure currently holds a `Comments`, where `Comments` is defined by:
```
data Comments = CommentNone
              | CommentFunc
```
Currently this option is just ignored when actually generating the code. My thinking is that a user should be able to decide whether they want functions, classes, or modules to be commented, or any combination thereof. So I would update `Choices` to take a list of `Comments`. Then define `Comments` as:
```
data Comments = CommentFunc | CommentClass | CommentModule
```
Where the `CommentNone` option is removed because it would simply be represented as the empty list. Then, any time we are generating a function, class, or module, we check whether the corresponding `Comments` data constructor is present in the list in the `Choices` structure, and generate a block comment if so. 

Another option is to remove `Comments` from the `Choices` structure altogether. With it in `Choices`, it is a "global" choice, which means the user can't decide to include comments only on one module that they feel warrants an explanation, and not others. To give them this flexibility, we could redefine `Comments` as:
```
data Comments = NoComments | WithComments
```
and then parameterize the `Func` and `Mod` data types in Drasil with a `Comments`, which determines whether that specific function/module should be headed by a block comment, allowing the comment decision to be made on the basis of individual functions/modules. There currently isn't a data type in Drasil to represent classes, but perhaps it makes sense to add one (it seems like something doesn't exist only because we haven't needed to add it yet).

**Comment by smiths** [comment]
@bmaclach your proposal sounds good to me, as long as it doesn't add too much complexity.  I like the level of control you are proposing, but I'm not sure there are many situations where a fine level of control on the presence or absence of comments is needed.  In my opinion, turning them "on" or "off" would cover most practical situations that come up.  

In your first proposal (`data Comments = CommentFunc | CommentClass | CommentModule`), I am confused by the difference between modules, functions and classes.  I'm assuming a module can contain classes and functions.  Does commenting the module automatically imply that its classes and functions will be commented, or is commenting a module related to the header information at the top of the file?

**Comment by bmaclach** [comment]
For commenting a module, I'm referring to the header information at the top of the file. You would be able to comment the module without commenting the classes and functions it contains (i.e. no header comment at the top of each class or function)

**Comment by smiths** [comment]
Thank you for the clarification @bmaclach.  I'm still not sure we need this level of control on comments, but as long as it won't take too much time, it has the advantage of providing future flexibility.  Turning everything "on" or "off" is still possible with your proposal, and other options are available too, just in case they are needed down the road.

**Comment by smiths** [comment]
When we generate comments, it would be great if they could be in doxygen format.  We could then generate nice code documentation without too much additional effort.  I would imagine that doxygen generation could be worked into the CI process that @Mornix and @danscime are working on.

**Comment by bmaclach** [comment]
@smiths Just to confirm: You're in support of the first of the two proposals laid out in my original comment?

**Comment by bmaclach** [comment]
Generating Doxygen-style comments shouldn't be a problem. 

Complexity-wise, the first option would be much simpler to implement given what we have now (or what we will have once #1435 gets merged in). The second option would be harder to implement because not all modules/functions ever exist as a `Mod` or `Func`, so it would require a bigger change to how we approach code generation (to make every module/function into `Mod`s and `Func`s before translating to GOOL code).

**Comment by smiths** [comment]
Yes @bmaclach, of the two proposals, I think the first is more useful.  I don't see the need to control the comments for individual modules.  Of course, there is a third proposal where you combine your two proposals so that you can customize the comments for each module.  I think that gets to the point of too high a granularity.  😄

**Comment by smiths** [comment]
@bmaclach, I'm glad to hear that my preference (option 1) is the easier one to implement.  I think doxygen-style comments would give us something that would be easy to "brag" about and show off.

**Comment by JacquesCarette** [comment]
This all sounds great. And definitely the kind of feature we want.

**Comment by bmaclach** [comment]
Closed by #1696.
